content/various/FastInput.h

```diff
@@ -0,0 +1,33 @@
+/**
+ * Author: chilli
+ * License: CC0
+ * Source:
+ * Description: Reads in arbitrary amounts of integral values.
+ * Status: tested on SPOJ INTEST
+ * Time: About 3x as fast as cin.
+ * Usage: int a, b[5]; ll c;
+ * read_int(a, b[0], c);
+ */
+#pragma once
+
+struct GC {
+    char buf[1 << 16 | 1];
+    int bc = 0, be = 0;
+    char operator()() {
+        if (bc >= be) {
+            be = fread(buf, 1, sizeof(buf) - 1, stdin);
+            buf[be] = bc = 0;
+        }
+        return buf[bc++];
+    }
+} gc;
+void read_int() {}
+template <class T, class... S>
+inline void read_int(T &a, S &... b) {
+	char c, s = 1;
+	while (isspace(c = gc()));
+	if (c == '-') s = -1, c = gc();
+	for (a = c - '0'; isdigit(c = gc()); a = a * 10 + c - '0');
+	a *= s;
+	read_int(b...);
+}
```

Comment:
I would prefer the API `int read_int();`, mostly because the implementation becomes shorter. (Let's you do `if (c == '-') return -read_int();`, e.g.) It also helps the compiler understand that pointer to the return value doesn't escape, but you've added inline so that's not really an argument.

It might be a good idea to replace `isspace(x)` by `x < 40` and similarly for `isdigit`, since those functions use the current locale and may be slow (?).

Reply:
I think one of the benefits of doing it this way is that we can read long longs/ints depending solely on what we pass in as arguments.

Lemme think about it a bit more. As book code, it doesn't matter as much how ergonomic it is, but I think it's still nice.

Reply:
Interestingly, replacing `isspace` with the boolean comparison makes a substantial difference (~20%). Replacing `isdigit` with the boolean comparison doesn't make much of a difference though.